gtsam/hybrid/tests/testDCSAM.cpp

```diff
@@ -0,0 +1,685 @@
+ of
+ * a Gaussian with large variance and an "alternative hypothesis" consisting
+ of
+ * a Gaussian with smaller variance. After initializing the continuous
+ variable
+ * far away from the ground-truth solution (x1 = 0), the discrete hypothesis
+ * selector will initially choose the null hypothesis.
+ *
+ * After a step of continuous optimization, the continuous solution will move
+ to
+ * x1 = 0 (since the problem is linear, we have convergence in one step).
+ * Finally, updating the discrete value will show that the correct discrete
+ * hypothesis will shift to the "alternative hypothesis."
+ */
+TEST(DCSAM, ContinuousMixture) {
+  using namespace discrete_mixture_fixture;
+
+  // Make an empty hybrid factor graph
+  HybridNonlinearFactorGraph dcfg;
+
+  // Make a symbol for a single continuous variable and add to KeyVector
+  KeyVector keys;
+  keys.push_back(x1);
+
+  std::vector<NonlinearFactorValuePair> factorComponents{
+      {f1, prior_noise1->negLogConstant()},
+      {fNullHypo, prior_noiseNullHypo->negLogConstant()}};
+
+  HybridNonlinearFactor dcMixture(dk, factorComponents);
+  dcfg.push_back(dcMixture);
+
+  // Test calculation of negative log probability
+  DiscreteValues dv1, dvNH;
+  dv1[dk.first] = 0;
+  dvNH[dk.first] = 1;
+  Values xvals;
+  xvals.insert(x1, 0.0);
+  const double negLogProb_1 = dcMixture.error(xvals, dv1);
+  const double negLogProb_NH = dcMixture.error(xvals, dvNH);
+
+  // As calculated in MATLAB using -log(normpdf(0,0,1)), -log(normpdf(0,0,8))
+  EXPECT_DOUBLES_EQUAL(0.9189, negLogProb_1, 1e-3);
+  EXPECT_DOUBLES_EQUAL(2.9984, negLogProb_NH, 1e-3);
+
+  // Let's make an initial guess for the continuous values
+  Values values;
+  double initVal = -2.5;
+  values.insert(x1, initVal);
+
+  // We also need an initial guess for the discrete variables (this will only be
+  // used if it is needed by your factors), here it is ignored.
+  DiscreteValues initialGuessDiscrete;
+  initialGuessDiscrete[dk.first] = 0;
+
+  // Let's make some factor graphs
+  DiscreteFactorGraph dfg;
+  NonlinearFactorGraph nfg;
+
+  // Create DiscreteFactor from initial guess
+  auto calculateError =
+      [&values](const NonlinearFactorValuePair &pair) -> double {
+    if (pair.first) {
+      auto gaussianNoiseModel = std::dynamic_pointer_cast<noiseModel::Gaussian>(
+          pair.first->noiseModel());
+      // `error` has the following contributions:
+      // - the scalar is the sum of all mode-dependent constants
+      // - factor->error(initial) is the error on the initial values
+      // - negLogK is log normalization constant from the noise model
+      return pair.second + pair.first->error(values);
+    } else {
+      // If the factor has been pruned, return infinite error
+      return std::numeric_limits<double>::infinity();
+    }
+  };
+  AlgebraicDecisionTree<Key> errors(dcMixture.factors(), calculateError);
+  DecisionTreeFactor dtf =
+      DiscreteFactorFromErrors(dcMixture.discreteKeys(), errors);
```

Comment:
I think this is recycling some of the code used in `DiscreteBoundaryFactor`, I wonder if you could just use that directly instead? Or maybe I'm misunderstanding something here.